### ISSUE TITLE
模型 loss 在训练的时候有时候会完全不变

### ISSUE THREAD

**Issue by Sky-zzt** [PaddlePaddle/Paddle#53354]
模型 loss 在训练的时候有时候会完全不变
paddle 1.8 
```
import paddle.fluid as fluid
import paddle
import numpy as np #引入数学计算的包
x=fluid.layers.data(name='x',shape=[-1,1],dtype='float32')
y = fluid.layers.data(name='y', shape=[1], dtype='float32')
hidden=fluid.layers.fc(input=x,size=2,act='relu')
net = fluid.layers.fc(input=hidden,size=1,act='tanh')
cost = fluid.layers.square_error_cost(input=net, label=y)
avg_cost = fluid.layers.mean(cost)
optimizer = fluid.optimizer.SGDOptimizer(learning_rate=0.01)
print(net.stop_gradient)
ones_label = fluid.layers.ones((fluid.layers.shape(y)[0],fluid.layers.shape(y)[1]),dtype='float32')
zeros_label = fluid.layers.zeros((fluid.layers.shape(y)[0],fluid.layers.shape(y)[1]),dtype='float32')
mask_label_one = fluid.layers.equal(ones_label,y)
mask_label_zero = fluid.layers.equal(zeros_label,y)
one_nonzero_label = fluid.layers.nonzero(mask_label_one,as_tuple=True)[0]
zero_nonzero_label = fluid.layers.nonzero(mask_label_zero,as_tuple=True)[0]
pos = fluid.layers.index_select(net, one_nonzero_label)
neg = fluid.layers.index_select(net, zero_nonzero_label)
pos = fluid.layers.reshape(pos,shape=[fluid.layers.shape(pos)[0],1])
pos_out = fluid.layers.expand(pos, expand_times=[fluid.layers.shape(pos)[0], fluid.layers.shape(neg)[0]])
new_t = fluid.layers.log(fluid.layers.sigmoid(fluid.layers.clip(pos_out-fluid.layers.t(neg),min=-15,max=15)))
new_t = fluid.layers.clip(new_t,min=-15,max=15)
print(ones_label.stop_gradient,',',mask_label_zero.stop_gradient,pos.stop_gradient)

avg_cost=-fluid.layers.mean(new_t)
opts = optimizer.minimize(avg_cost)
place = fluid.CPUPlace()
exe = fluid.Executor(place)

exe.run(fluid.default_startup_program())
x_data = np.array([[1.], [2.0], [3.0], [4.0], [5.0], [6.0],[7.0]]).astype('float32')
y_data = np.array([[0.], [0.0], [0.0], [0.0], [1.0], [1.0],[01.0]]).astype('float32')
for pass_id in range(50):
    train_cost,nonzer = exe.run(program=fluid.default_main_program(),
                        feed={'x': x_data, 'y': y_data},
                        fetch_list=[avg_cost,zero_nonzero_label])
    if pass_id% 3 ==0 :print("Pass:%d, Cost:%0.5f" % (pass_id, train_cost[0]))

```

![image](https://user-images.githubusercontent.com/52705155/234464997-2a8153fe-389f-4c21-8312-6b87d43d6d9e.png)

![image](https://user-images.githubusercontent.com/52705155/234465685-3de2efe3-6360-4992-9117-30305940d6ea.png)

**Comment by kangguangli** [comment]
你好，请问使用的paddle版本是什么呢？

**Comment by Sky-zzt** [comment]
> 你好，请问使用的paddle版本是什么呢？

1.8

**Comment by kangguangli** [comment]
你好，推荐试试使用2.4版本或者develop版本看下这个问题是否依旧存在。
从1.8到2.0版本，如果需要升级，可以参考[升级指南](https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/model_convert/update_cn.html#shengjigaiyao)

**Comment by kangguangli** [comment]
顺便麻烦问下，你是有什么特殊场景的需求，必须使用paddle1.8版本吗？

**Comment by Sky-zzt** [comment]
> 你好，推荐试试使用2.4版本或者develop版本看下这个问题是否依旧存在。 从1.8到2.0版本，如果需要升级，可以参考[升级指南](https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/model_convert/update_cn.html#shengjigaiyao)

```
import paddle.fluid as fluid
import paddle
import numpy as np
paddle.enable_static()
x=fluid.layers.data(name='x',shape=[-1,1],dtype='float32')
y = fluid.layers.data(name='y', shape=[1], dtype='float32')
hidden=fluid.layers.fc(input=x,size=2,act='relu')
net = fluid.layers.fc(input=hidden,size=1,act='tanh')
cost = fluid.layers.square_error_cost(input=net, label=y)
avg_cost = fluid.layers.mean(cost)
optimizer = fluid.optimizer.SGDOptimizer(learning_rate=0.01)
print(net.stop_gradient)
sig = paddle.nn.Sigmoid()
ones_label = fluid.layers.ones((fluid.layers.shape(y)[0],fluid.layers.shape(y)[1]),dtype='float32')
zeros_label = fluid.layers.zeros((fluid.layers.shape(y)[0],fluid.layers.shape(y)[1]),dtype='float32')
mask_label_one = fluid.layers.equal(ones_label,y)
mask_label_zero = fluid.layers.equal(zeros_label,y)
pos = paddle.masked_select(net,mask_label_one)
neg = paddle.masked_select(net,mask_label_zero)
pos = pos.reshape([pos.shape[0],1])
neg = neg.reshape([neg.shape[0],1])
pos_out = paddle.expand(pos, shape=[pos.shape[0], neg.shape[0]])
new_t = paddle.log(sig(pos_out-neg.T))
avg_cost=-paddle.mean(new_t)
opts = optimizer.minimize(avg_cost)
place = fluid.CPUPlace()
exe = fluid.Executor(place)

exe.run(fluid.default_startup_program())
x_data = np.array([[1.], [2.0], [3.0], [4.0], [5.0], [6.0],[7.0]]).astype('float32')
y_data = np.array([[0.], [0.0], [0.0], [0.0], [1.0], [1.0],[01.0]]).astype('float32')
for pass_id in range(50):
    train_cost = exe.run(program=fluid.default_main_program(),
                        feed={'x': x_data, 'y': y_data},
                        fetch_list=[avg_cost])
    if pass_id% 3 ==0 :print("Pass:%d, Cost:%0.5f" % (pass_id, train_cost[0]))


```

这个2.4 版本代码 执行多次后也会复现
![image](https://user-images.githubusercontent.com/52705155/234498409-e1b3b470-973b-453d-af2d-58409dea5203.png)

**Comment by kangguangli** [comment]
你好，经过我的调查，问题应该是这样的，如果在某种初始化设置下，这里net的输出值是一样的，那么后续计算得到的`pos_out-neg.T`是一个零向量，因而计算得到的loss就会固定在`log1/2=-0.6931471805599453`。
在梯度更新时，此时梯度依然会正常更新，但是更新梯度时，`pos_out-neg.T`是一个零向量，net@grad也是 net的输出值依然完全一致，打印出这里net的值会发现它在变化，但是矩阵的梯度为0。这里的原因是第一层fc使用了relu激活，如果恰巧第一层fc结束后得到的全是负数，那么relu会把它变成0，进而反向计算就会得到这样的结果。
我为上面的程序增加了中间结果的输出，你可以运行观察一下：
```
import paddle.fluid as fluid
import paddle
import numpy as np
paddle.enable_static()
x=fluid.layers.data(name='x',shape=[-1,1],dtype='float32')
y = fluid.layers.data(name='y', shape=[1], dtype='float32')
hidden=paddle.static.nn.fc(x=x,size=2,activation='relu')
net = paddle.static.nn.fc(x=hidden,size=1,activation='tanh')
optimizer = paddle.optimizer.SGD(learning_rate=0.01)
print(net.stop_gradient)

sig = paddle.nn.Sigmoid()
ones_label = fluid.layers.ones((fluid.layers.shape(y)[0],fluid.layers.shape(y)[1]),dtype='float32')
zeros_label = fluid.layers.zeros((fluid.layers.shape(y)[0],fluid.layers.shape(y)[1]),dtype='float32')
mask_label_one = fluid.layers.equal(ones_label,y)
mask_label_zero = fluid.layers.equal(zeros_label,y)

pos = paddle.masked_select(net,mask_label_one)
neg = paddle.masked_select(net,mask_label_zero)

pos = pos.reshape([pos.shape[0],1])
neg = neg.reshape([neg.shape[0],1])
pos_out = paddle.expand(pos, shape=[pos.shape[0], neg.shape[0]])
new_t = paddle.log(sig(pos_out-neg.T))
avg_cost=-paddle.mean(new_t)
opts = optimizer.minimize(avg_cost)
place = fluid.CPUPlace()
exe = fluid.Executor(place)

print(fluid.default_main_program())

exe.run(fluid.default_startup_program())
x_data = np.array([[1.], [2.0], [3.0], [4.0], [5.0], [6.0],[7.0]]).astype('float32')
y_data = np.array([[0.], [1.0], [0.0], [0.0], [1.0], [1.0],[01.0]]).astype('float32')
for pass_id in range(50):
    train_cost = exe.run(program=fluid.default_main_program(),
                        feed={'x': x_data, 'y': y_data},
                        fetch_list=[avg_cost, "fc_1.tmp_2", 
                        "masked_select_0.tmp_0", 
                        "masked_select_1.tmp_0",
                        "fc_1.tmp_2@GRAD",
                        "fc_0.w_0", 
                        "fc_1.w_0", 
                        "fc_0.w_0@GRAD", 
                        "fc_1.w_0@GRAD",
                        "fc_0.tmp_2",
                        "fc_1.tmp_0@GRAD"])
    print("loss:", train_cost[0])
    print("net:", train_cost[1])
    print("paddle.masked_select(net,mask_label_one): ", train_cost[2])
    print("paddle.masked_select(net,mask_label_zero):", train_cost[3])
    print("net grad", train_cost[4])
    print("fc_0.w_0", train_cost[5])
    print("fc_1.w_0", train_cost[6])
    print("fc_0.w_0@GRAD", train_cost[7])
    print("fc_1.w_0@GRAD", train_cost[8])
    print("fc_0.tmp_2", train_cost[8])
    print("fc_1.tmp_0@GRAD", train_cost[8])
```
